### PR TITLE
[core][dashboard] push down job_or_submission_id to GCS.

### PR DESCRIPTION
**Pull request by rynewang**
GCS API GetAllJobInfo serves Dashboard APIs, even for only 1 job. This becomes slow when the number of jobs are high. This PR pushes down the job filter to GCS to save Dashboard workload.

This API is kind of strange because the filter `job_or_submission_id` is actually Either a Job ID Or a job_submission_id. We don't have an index on the latter, and some jobs don't have one. So we still GetAll from Redis; and filter by both IDs after that and before doing more RPC calls.

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 1]
```diff
@@ -145,6 +145,16 @@ void GcsJobManager::AddJobFinishedListener(JobFinishListenerCallback listener) {
 void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                                         rpc::GetAllJobInfoReply *reply,
                                         rpc::SendReplyCallback send_reply_callback) {
+  // Get all job info. This is a complex operation:
+  // 1. One GetAll from the job table.
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:149]: I think https://github.com/ray-project/ray/pull/46861 will help here as well

Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:149]: surprised I haven't merged that; will debug now

File: python/ray/_private/gcs_aio_client.py [hunk 2]
```diff
@@ -200,9 +200,9 @@ async def internal_kv_keys(
         return await self._async_proxy.internal_kv_keys(prefix, namespace, timeout)
 
     async def get_all_job_info(
-        self, timeout: Optional[float] = None
+        self, job_or_submission_id: str = None, timeout: Optional[float] = None
```
Comment (jjyao) [inline python/ray/_private/gcs_aio_client.py:203]:
```suggestion
        self, job_or_submission_id: Optional[str] = None, timeout: Optional[float] = None
```

**Review by jjyao — COMMENTED** [review]
Which Dashboard endpoint will benefit from this?

**Comment by rynewang** [comment]
> Which Dashboard endpoint will benefit from this?

```
@routes.get("/api/jobs/{job_or_submission_id}")
```

@alexeykudinkin can you confirm this is a hot API we encountered?

**Comment by liuxsh9** [comment]
Maybe #47530 can also give a performance boost? Would love to hear your thoughts!

File: src/ray/protobuf/gcs_service.proto [hunk 3]
```diff
@@ -39,6 +39,9 @@ message MarkJobFinishedReply {
 message GetAllJobInfoRequest {
   // The number of jobs to return. If not specified, return all jobs.
   optional int32 limit = 1;
+  // If set, only return the job with that job id in hex, or the job with that
+  // job_submission_id.
+  optional bytes job_or_submission_id = 3;
```
Comment (alexeykudinkin) [inline src/ray/protobuf/gcs_service.proto:44]: Let's make this repeated to make this proper batch API

File: src/ray/gcs/gcs_server/gcs_job_manager.cc [hunk 4]
```diff
@@ -145,6 +145,16 @@ void GcsJobManager::AddJobFinishedListener(JobFinishListenerCallback listener) {
...
+  // 2. For each job, Send RPC to core work for is_running_tasks value.
```
Comment (alexeykudinkin) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:150]:
```suggestion
  // 2. For each job, Send RPC to core worker for is_running_tasks value.
```

File: src/ray/protobuf/gcs_service.proto (hunk 3 above)
Comment (rynewang) [reply src/ray/protobuf/gcs_service.proto:44]: I would refrain until we find a need for it

File: python/ray/dashboard/modules/job/utils.py [hunk 5]
```diff
@@ -148,7 +148,9 @@ async def parse_and_validate_request(
 
 
 async def get_driver_jobs(
-    gcs_aio_client: GcsAioClient, timeout: Optional[int] = None
+    gcs_aio_client: GcsAioClient,
+    job_or_submission_id: str = None,
```
Comment (jjyao) [inline python/ray/dashboard/modules/job/utils.py:152]:
```suggestion
    job_or_submission_id: Optional[str] = None,
```

File: python/ray/dashboard/modules/job/utils.py [hunk 6]
```diff
@@ -157,7 +159,9 @@ async def get_driver_jobs(
     It's keyed by the submission job's submission id.
     Only the last driver of a submission job is returned.
     """
```
Comment (jjyao) [inline python/ray/dashboard/modules/job/utils.py:161]: Update the docstring to describe `job_or_submission_id`

File: src/ray/gcs/gcs_client/accessor.h [hunk 7]
```diff
@@ -247,17 +247,21 @@ class JobInfoAccessor {
...
 
   /// Get all job info from GCS synchronously.
   ///
-  /// \param timeout_ms -1 means infinite.
+  /// \param job_or_submission_id If not empty, filter the jobs with this id.
   /// \param[out] job_data_list The list of job data retrieved from GCS.
+  /// \param timeout_ms -1 means infinite.
   /// \return Status
-  virtual Status GetAll(std::vector<rpc::JobTableData> &job_data_list,
+  virtual Status GetAll(const std::string &job_or_submission_id,
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/accessor.h:263]: nit: using std::optional is more explicit

Comment (jjyao) [inline src/ray/gcs/gcs_client/accessor.h:263]: nit: using std::optional is more explicit

**Review by jjyao — COMMENTED** [review]
lg

**Comment by jjyao** [comment]
rebase to fix rllib failure

**Review by jjyao — APPROVED** [review]

**Pull request merged**